test/ui/cypress/integration/serving.spec.js

```diff
@@ -0,0 +1,168 @@
+describe('OCP UI for Serverless', () => {
+
+  class ShowcaseKservice {
+    constructor(ops = {}) {
+      this.app = ops.app || 'demoapp'
+      this.name = ops.name || 'showcase'
+      this.namespace = ops.namespace || Cypress.env('TEST_NAMESPACE')
+      this.image = ops.image || {
+        regular: 'quay.io/cardil/knative-serving-showcase:2-send-event',
+        updated: 'quay.io/cardil/knative-serving-showcase-js'
+      }
+    }
+
+    makeRequest() {
```

Comment:
> Maybe showServiceDetails could be called at the beginning of these functions to ensure that?
That would lengthen the whole test. Also, in traffic distribution test I call 8 times the request so to make sure both revisions get traffic. Adding `showServiceDetails` will make to sloooow.

Reply:
The makeRequest function does multiple things which are not clear from the function name:
* get the right link in UI for the HTTP request
* call the HTTP request
* verify things specific for the knative-serving-showcase
I think it will become clear later that calling this HTTP request will be useful in other tests as well so we will want to extract this part to a different class. Also, instead of doing the checks at the end it could also return the response and it could be verified/checked in the test itself and not in the function that sends the request.

Reply:
My view of ShowcaseKservice is this:
Functions that belong here are checkScale, showServiceDetails, getAppURL
Functions that could be extracted and made reusable: 
* makeRequest (just calling the request and returning the response, not doing checks, not getting the right URL/link)
* deployImage - this sounds useful for future different tests as well
* removeApp - same as above
The ShowcaseKservice could then just have thin wrappers that pass the right object to the functions above.
This way the ShowcaseKservice would be more focused on the given use case, and there would be reusable parts ready for other tests. Anyway, these are things that will probably become apparent later when we develop more tests.

Reply:
It can't return anything. It's async. It at best, could return a promise of response. But, then what's the difference with built-in `cy.request` function?

Reply:
Hmm. I'm not very good at javascript but this function seems to be doing more then it should, according to its name. Is it possible to get the app URL outside that function and pass it to it? Something like `showcaseKsvc.getUrl().then((url) => checkResponse(url))` ? This could reduce remove getting the URL when called in the loop (8 times).

Reply:
That's possible. I'll try doing that.

Reply:
@cardil will you try to do that or this is ready for integrating?

Reply:
This is done.

Reply:
Oh. I didn't notice. Thanks:)